File: LeetCode/Hard/3474-lexicographically-smallest-generated-string/3474-lexicographically-smallest-generated-string.py
class Solution:
    def generateString(self, str1: str, str2: str) -> str:
        len_str2 = len(str2)
        result = ['a'] * (len(str1) + len_str2 - 1)
        committed = set()

        for i, c in enumerate(str1):
            if c == 'F':
                continue
            for j in range(len_str2):
                w = i + j
                if w in committed and result[w] != str2[j]:
                    return ""
                result[w] = str2[j]
                committed.add(w)
        
        str2_list = list(str2)
        for i, c in enumerate(str1):
            if c == 'T' or result[i:i + len_str2] != str2_list:
                continue
            
            for j in range(len_str2 - 1, -1, -1):
                rightmost_idx = i + j
                if not rightmost_idx in committed:
                    result[rightmost_idx] = 'b'
                    committed.add(rightmost_idx)
                    break
            else:
                return ""
        
        return "".join(result)

File: LeetCode/Hard/3474-lexicographically-smallest-generated-string/3474-lexicographically-smallest-generated-string.cpp
class Solution {
public:
    string generateString(string str1, string str2) {
        string result(str1.size() + str2.size() - 1, 'a');
        vector<bool> fixed(str1.size() + str2.size() - 1);

        for (auto [i, condition] : views::enumerate(str1))
        {
            if (condition == 'F') continue;

            for (auto j = 0; j < str2.size(); ++j)
            {
                auto w = i + j;
                auto c_pattern = str2[j];

                if (fixed[w] && result[w] != c_pattern) return "";

                result[w] = c_pattern;
                fixed[w] = true;
            }
        }

        string_view pattern = str2;
        for (auto [i, condition] : views::enumerate(str1))
        {
            if (condition == 'T' || string_view(result.data() + i, str2.size()) != pattern) continue;

            auto j = str2.size();
            bool wasChanged = false;
            for (; j-- > 0;)
            {
                auto w = i + j;
                cout << w << endl;
                if (fixed[w]) continue;

                result[w] = 'b';
                fixed[w] = wasChanged = true;
                break;
            }

            if (!wasChanged) return "";
        }

        return result; 
    }
};